src/groth16/prover.rs

```diff
@@ -254,13 +272,29 @@ where
         info!("GPU Multiexp is NOT supported!");
     }
 
-    let h = multiexp(
-        &worker,
-        params.get_h(a.len())?,
-        FullDensity,
-        a,
-        &mut multiexp_kern,
-    );
+    let mut keep_cpu = false;
+
+    let h = if !check_for_higher_prio!() || keep_cpu {
+        #[cfg(feature = "gpu")]
+        {
+            info!("Multiexp 1 Prover found acquire lock, switching to CPU");
+            // Free the incoming process to use the GPU
+            if !keep_cpu {
+                keep_cpu = true;
+                lock.unlock()?;
+            }
+        }
+        multiexp(&worker, params.get_h(a.len())?, FullDensity, a, &mut None)
+    } else {
+        info!("Multiexp 1 Prover NO acquire lock, keeping GPU");
+        multiexp(
+            &worker,
+            params.get_h(a.len())?,
+            FullDensity,
+            a,
+            &mut multiexp_kern,
+        )
+    };
```

Comment:
why not add this logic to `fn multiexp` instead, that way you don't have to duplicate it in here